Omit key in output if required is false and value is none

ahankinson opened the pull request:
In Django REST Framework, if a field is set to required=False and the value is None, the key will be omitted from the output.

Currently in serpy, if required is set to False and a value is null, the output will contain a null value for that key.

This commit changes the behaviour to match DRF. If a value is None and a key is set to required=False, the key will not appear in the output.

Includes an updated unit test to reflect this behaviour. A test case for type coercion on None values was also created to better reflect the code under test.

ahankinson committed:
Omit key in output if required is false and value is none

In Django REST Framework, if a field is set to required=False and the value is None, the key will be omitted from the output.

Currently in serpy, if required is set to False and a value is null, the output will contain a null value for that key.

This commit changes the behaviour to match DRF. If a value is None and a key is set to required=False, the key will not appear in the output.

Includes an updated unit test to reflect this behaviour. A test case for type coercion on None values was also created to better reflect the code under test.

jscn commented:
I was just now looking at the docs trying to see if this was possible, so I'd love to see this merged.

@ahankinson looks like you've got a couple PEP8 violations.

ahankinson committed:
Fix PEP8 violations

ahankinson commented:
Stupidest PEP8 violations ever. Thanks.

clarkduvall commented on serpy/serializer.py:
```diff
@@ -112,7 +112,13 @@ def _serialize(self, instance, fields):
+
+            if required:
+                v[name] = result
+            elif not required and result:
```
so there's a problem here if the result is a falsy value, like 0 or '' it won't be in the serialized dict. I think this would be more correct as `result is not None`

You can combine the checks as:
```py
if required or result is not None:
    v[name] = result
```
The `else: continue` is also unnecessary since it's at the end of the loop anyway.

It would be great if you could add a test for the case where the field is not required but produces a falsy value like 0.

clarkduvall commented:
@ahankinson is this  behavior specified in the DRF docs anywhere? I can't seem to find it. It seems like the key is omitted from the output not if the value is None but if the corresponding key was not present in the input. For example:

```py
from rest_framework import serializers

class Foo(serializers.Serializer):
    foo = serializers.ReadOnlyField(required=False)

>>> Foo({'foo': None}).data
{'foo': None}
>>> Foo({}).data
{}
```
To match that behavior might be a little more difficult given how serpy is currently designed.

ahankinson commented:
The behaviour is documented here: http://www.django-rest-framework.org/api-guide/fields/#required

```
Normally an error will be raised if a field is not supplied during deserialization. Set to false if this field is not required to be present during deserialization.

Setting this to False also allows the object attribute or dictionary key to be omitted from output when serializing the instance. If the key is not present it will simply not be included in the output representation.

Defaults to True.
```

ahankinson committed:
Updated tests to check for falsy values

This commit adds a test to check the behaviour of falsy values and its interactions with the 'required' flag on fields.

ahankinson committed:
Simplify checks for required fields

This commit simplifies the logic for checks to output required fields.

ahankinson commented:
In addition to the fixes I also wrote a quick unit test to mirror the falsy values one I just added, but for Django REST Framework. I didn't commit it since I didn't want to make DRF and Django a dependency, but if you wanted to confirm that the behaviour is the same, here it is:

```python
from django.conf import settings
settings.configure()

import django
django.setup()

from .obj import Obj
from rest_framework import serializers
import unittest


class TestDRFCompat(unittest.TestCase):
    def test_behaviour_with_falsy_values(self):
        class ASerializer(serializers.Serializer):
            a = serializers.IntegerField(required=False)
            b = serializers.BooleanField(required=False)
            c = serializers.CharField()
            d = serializers.CharField(required=False)

        o = Obj(a=0, b=False, c="", d="")
        data = ASerializer(o).data
        self.assertTrue('a' in data)
        self.assertTrue('b' in data)
        self.assertTrue('c' in data)
        self.assertTrue('d' in data)
```

ahankinson commented:
Bleh.

After digging around, it appears you are right, @clarkduvall. I can't seem to reproduce this behaviour with DRF. I'm feeling like I'm going crazy, though, since I could have sworn that it was the case that None values were omitted from the output.

I'm going to keep looking.

ahankinson committed:
Allow missing keys on objects

Previously, a missing key on an object defined in the serializer would raise an attribute exception when serpy tried to access the attribute, even if the required=False flag was set.

In Django REST Framework, a missing key on a non-required field would result in that field not being reflected in the output.

This commit provides compatibility with DRF behaviour in serpy. Now, if an attribute is not present on the object to be serialized, and that field is not required, it will not be present in the reflected output.

This commit also includes improved unit tests for these cases, including tests for both serpy's Serializer and DictSerializer classes in the case of required values.

ahankinson committed:
Formatting: PEP8

ahankinson commented:
OK, the latest commit seems to reflect DRF's behaviour better. Now when a key is omitted from the object to be serialized, and it is not a required field, serpy will not include it in the output. If it is a required field (default) it will raise a KeyError / AttributeError.

Here's an updated unit test to confirm DRF's behaviour:

```python
from django.conf import settings
settings.configure()

import django
django.setup()

from .obj import Obj
from rest_framework import serializers
import unittest


class TestDRFCompat(unittest.TestCase):
    def test_optional_fields(self):
        class ASerializer(serializers.Serializer):
            a = serializers.IntegerField(required=False)
            b = serializers.CharField(required=False)

        class BSerializer(serializers.Serializer):
            a = serializers.IntegerField()

        o = Obj()
        data = ASerializer(o).data
        self.assertTrue('a' not in data)
        self.assertTrue('b' not in data)

        o = Obj()
        with self.assertRaises(AttributeError) as context:
            BSerializer(o).data

        o = Obj(a='5')
        data = ASerializer(o).data
        self.assertEqual(data['a'], 5)

    def test_behaviour_with_falsy_values(self):
        class ASerializer(serializers.Serializer):
            a = serializers.IntegerField(required=False)
            b = serializers.BooleanField(required=False)
            c = serializers.CharField()
            d = serializers.CharField(required=False)

        o = Obj(a=0, b=False, c="", d="")
        data = ASerializer(o).data
        self.assertTrue('a' in data)
        self.assertTrue('b' in data)
        self.assertTrue('c' in data)
        self.assertTrue('d' in data)
```

ahankinson committed:
Improve testing of nested serializers

mheisig commented:
One alternative would be the way Marshmallow handles this with an `exclude` parameter. Either way, I'd find this very useful.

ahankinson commented:
@clarkduvall would it be acceptable to add a field (perhaps `omit=True|False`?) to control this behaviour?

belljustin committed:
removes context manager from test

anentropic commented:
TBH I think the `required` arg should be renamed as it does not do anything like the familiar arg of same name in DRF/Django... something like `always_coerce` to show that its actual effect is on whether or not type-coercion happens.

I already found myself adding `required=False` to all my fields to avoid inconsistent behaviour in case of `None` values, e.g. `StrField` will return `"None"` while `IntField` raises a `TypeError`.  I feel like `required=False` should be the default, least surprising.

So I also agree that the behaviour desired in this particular PR should be via a differently named arg, such as `omit` or `exclude`.

clarkduvall commented:
@ahankinson  thanks for bringing up this issue, I ended up merging #38. I'm going to close this, please reopen if #38 did not solve the issue.

anentropic commented:
I am happy with the behaviour merged in #38 

However there is a slightly different and separate requirement in this issue, to have fields omitted from the response if value is null. I don't much like APIs which do this but sometimes we have to be compatible with them.

I have been using following workaround in my project:

```python
class OptionalFieldsSerializerMixin(object):
    """
    Tells the serializer to omit the specified fields from the response if
    they have a null value.

    e.g.

        class MySerializer(OptionalFieldsSerializerMixin, serpy.Serializer):

            _optional_fields = ('whatever',)
            _optional_fields_strict = ('whenever',)

            name = serpy.StrField(required=False)
            whatever = serpy.IntField(required=False)  # will be omitted if value is 'falsy'
            whenever = serpy.IntField(required=False)  # will be omitted if value is None

    """

    def _serialize(self, instance, fields):
        data = super(OptionalFieldsSerializerMixin, self)._serialize(instance, fields)
        if hasattr(self, '_optional_fields'):
            for key in self._optional_fields:
                if not data[key]:
                    del data[key]
        if hasattr(self, '_optional_fields_strict'):
            for key in self._optional_fields_strict:
                if data[key] is None:
                    del data[key]
        return data
```
if I was making a PR I would instead change something in base field class to add an `optional` attribute (this would mean we have both `required` and `optional` attrs, with the same meanings as they have in API Blueprint/MSON for example) instead of these serializer attrs.
